File: ch03/src/cpp/lazy_init_with_indirection.cpp
#include <atomic>
#include <chrono>
#include <cstdint>
#include <ctime>
#include <iostream>
#include <ostream>
#include <syncstream>
#include <thread>
#include <vector>

struct Data {
    std::int64_t a;
    std::int64_t b;
};

std::ostream& operator<<(std::ostream& os, const Data& d)
{
    os << "{" << d.a << ", " << d.b << "}";
    return os;
}

Data* generate_data()
{
    std::this_thread::sleep_for(std::chrono::milliseconds { 20 });
    return new Data(12, 34);
}

auto get_utc_time()
{
    using namespace std::chrono;
    static const auto kUtcZonePtr = locate_zone("UTC");
    const auto now = system_clock::now();
    return zoned_time<nanoseconds, const time_zone*> { kUtcZonePtr, now };
}

Data& get_data()
{
    static auto sPtr = std::atomic<Data*> { nullptr };

    auto p = sPtr.load(std::memory_order_acquire);

    const auto tid { std::this_thread::get_id() };
    auto scout = std::osyncstream { std::cout };
    if (p != nullptr) {
        // p is not nullptr. Some thread must have initialized sPtr with memory order release
        // (see the else clause below). The release-acquire order ensures that p points to
        // initialized data, so there's nothing more for this thread to do.
        scout << get_utc_time() << ": Thread " << tid << ": uses existing data\n";
    } else {
        scout << get_utc_time() << ": Thread " << tid << ": attempts to initialize data\n";

        // Initialize the data and store it in p
        p = generate_data();

        // Try to store p in sPtr
        Data* n = nullptr;
        if (sPtr.compare_exchange_strong(n, p, std::memory_order_release,
                std::memory_order_acquire)) {
            // The compare_exchange_strong has performed a read-modify-write operation.

            scout << get_utc_time() << ": >>>>>>>> Thread " << tid << " wins the race to initialize data!\n";
        } else {
            // The compare_exchange_strong has performed a load operation.

            scout << get_utc_time() << ": Thread " << tid << " loses the race\n";
            delete p;
            p = n;
        }
    }
    scout << get_utc_time() << ": " << tid << ": " << *p << " @ " << p << '\n';

    return *p;
}

int main()
{
    std::vector<std::jthread> handles;
    handles.reserve(100);
    for (auto i { 0 }; i < 100; ++i) {
        handles.emplace_back(get_data);
        std::this_thread::sleep_for(std::chrono::milliseconds { 1 });
    }
    return 0;
}

File: ch03/src/bin/lazy_init_with_indirection.rs
use std::{
    sync::atomic::{AtomicPtr, Ordering},
    thread,
    time::Duration,
};

struct Data {
    v: i32,
}

fn generate_data() -> Data {
    thread::sleep(Duration::from_millis(50));

    Data { v: 42 }
}

fn get_data() -> &'static Data {
    static PTR: AtomicPtr<Data> = AtomicPtr::new(std::ptr::null_mut());

    let mut p = PTR.load(Ordering::Acquire);

    if p.is_null() {
        println!(
            "Thread {:?} loads NULL, trying to initialize data",
            thread::current().id()
        );

        // Generate new data with generate_data
        // Store it in a new allocation with Box::new
        // Turn this box into a pointer with Box::into_raw,
        // so that we can attempt to store it in PTR
        p = Box::into_raw(Box::new(generate_data()));

        // Attempt to store p in PTR
        if let Err(e) = PTR.compare_exchange(
            std::ptr::null_mut(), // Check against std::ptr::null_mut()
            p,                    // If the same, then will set PTR to p
            Ordering::Release, // The ordering for the read-modify-write operation if the comparison above succeeds
            Ordering::Acquire, // The ordering for the load operation if the comparison above fails
        ) {
            println!(
                "Thread {:?} loses the race to initialize data.",
                thread::current().id()
            );

            // PTR was already set to e: Another thread won the race to initialize data.

            // The ordering is Acquire. Together with the Release ordering of the winning thread,
            // we can ensure that the initialization of the pointer e happens-before the
            // assignment of e to p below.
            // In other word, these orderings ensure that p points to initialized data.

            // Here we turn the raw pointer back into a box and drop it.
            drop(unsafe { Box::from_raw(p) });

            // Use the pointer from the thread that won the race.
            p = e;
        } else {
            println!(
                "==================== Thread {:?} did the initialization! ====================",
                thread::current().id()
            );

            // PTR has been atomically set to p
            // The ordering is Release
        }
    } else {
        // PTR already pointed to valid data
        println!("Thread {:?}: get existing data", thread::current().id());
    }

    unsafe { &*p }
}

fn main() {
    thread::scope(|s| {
        for _ in 0..100 {
            s.spawn(|| {
                let v = get_data().v;
                println!("Thread {:?}: data = {}", thread::current().id(), v);
            });

            thread::sleep(Duration::from_millis(5));
        }
    });
}
